workers/utils.py

```diff
@@ -319,3 +320,16 @@ def setup_environ() -> None:
     # Apply
     os.environb.clear()
     os.environb.update(env)
+
+FUZZER_CMD_PORT = 7055
+
+class _PausedFuzzers:
```

Comment:
We probably need some way to configure or detect whether fuzzer is
running. This is why back at the beginning I suggested using Unix
sockets for communications since then NayDuck worker could see if the
socket exists and don’t bother trying to pause if it doesn’t.

We probably would also like some kind of retry logic, i.e. try pausing
and unpausing a few times on failures.  It is communicating over
localhost so chances of network failures are zero but still maybe it’s
worth doing something like:

```python
last_exception = None
for n in range(3):
    try:
        requests.get(...).raise_for_status()
        break
    except requests.exceptions.RequestException as ex:
        last_exception = ex
        time.sleep(1.5**n)
else:
    action = 'pause' # 'unpause'
    print(f'Failed to {action} fuzzers: {last_exception}',
          file=sys.stderr)
```

Lastly, there’s one more theoretical race condition I kinda though not
really worry about.  If fuzzer starts after NayDuck starts running the
test, fuzzer would then go ahead and start doing its job.  Perhaps we
need some kind of lock file as well?

Reply:
I've implemented the backoff strategy you mentioned, and “solved” the race condition by adjusting the systemd unit file — now I think the only issue that could happen is if the fuzzer were to start before the worker but to take more than the maximum backoff duration to actually spin up the webserver, which would sound surprising to me.

Reply:
(A solution to avoid this problem would be to integrate the fuzzer with systemd more so systemd only considers it as having started up once the http server is up, but I'm not sure it's really worth it actually doing so)

Reply:
This doesn’t fully solve the issue. A fuzzer could be restarted while worker is running but it does seem like it would help.

Reply:
Right, but I think a fuzzer restart could happen only with manual intervention restarting the systemd service? (because the daily automated fuzzer restarts shouldn't happen when the fuzzers are paused)
And so if there's manual intervention it's the responsibility of the one actually doing the manual operation, as there's no way to guard against the user eg. just shutting down the worker :sweat_smile:

Reply:
The fuzzing service could crashs and be restarted by systemd. And technically speaking in extreme cases even if worker is started after the fuzzer, the fuzzer could take a while to start the HTTP server. We probably can live with this but strictly speaking this isn’t a full solution.

Reply:
You're right, I hadn't thought of a crash-induced restart. Let's come back to this on the first day there'll be issues then :)